File: test/Benchmark/Prepare.cs
using System;
using TDengine.Driver;
using NativeMethods = TDengine.Driver.Impl.NativeMethods.NativeMethods;

namespace Benchmark
{
    internal class Prepare
    {
        string Host { get; set; }
        ushort Port { get; set; }
        string Username { get; set; }
        string Password { get; set; }
        string db = "benchmark";

        readonly string createStb = "create table if not exists stb (ts timestamp" +
                                    ",bl bool" +
                                    ",i8 tinyint" +
                                    ",i16 smallint" +
                                    ",i32 int" +
                                    ",i64 bigint" +
                                    ",u8 tinyint unsigned" +
                                    ",u16 smallint unsigned" +
                                    ",u32 int unsigned" +
                                    ",u64 bigint unsigned" +
                                    ",f32 float" +
                                    ",d64 double" +
                                    ",bnr binary(50)" +
                                    ",nchr nchar(50))" +
                                    "tags(t_bl bool" +
                                    ",t_i8 tinyint" +
                                    ",t_i16 smallint" +
                                    ",t_i32 int" +
                                    ",t_i64 bigint" +
                                    ",t_u8 tinyint unsigned" +
                                    ",t_u16 smallint unsigned" +
                                    ",t_u32 int unsigned" +
                                    ",t_u64 bigint unsigned" +
                                    ",t_f32 float" +
                                    ",t_d64 double" +
                                    ",t_bnr binary(50)" +
                                    ",t_nchr nchar(50));";

        readonly string createJtb = "create table if not exists jtb(ts timestamp" +
                                    ", bl bool" +
                                    ", i8 tinyint" +
                                    ", i16 smallint" +
                                    ", i32 int" +
                                    ", i64 bigint" +
                                    ", u8 tinyint unsigned" +
                                    ", u16 smallint unsigned" +
                                    ", u32 int unsigned" +
                                    ", u64 bigint unsigned" +
                                    ", f32 float" +
                                    ", d64 double" +
                                    ", bnr binary(50)" +
                                    ",nchr nchar(50))" +
                                    "tags(json_tag json);";

        readonly string createStb1 = "create table if not exists stb_2 using stb tags(true" +
                                     ",-1" +
                                     ",-2" +
                                     ",-3" +
                                     ",-4" +
                                     ",1" +
                                     ",2" +
                                     ",3" +
                                     ",4" +
                                     ",3.1415" +
                                     ",3.14159265358979" +
                                     ",'bnr_tag_1'" +
                                     ",'ncr_tag_1');";

        readonly string createJtb1 =
            "create table if not exists jtb_2 using jtb tags('{\"jtag_bool\":false,\"jtag_num\":3.141592653,\"jtag_str\":\"beijing\",\"jtag_null\":null}');";

        public Prepare(string host, string userName, string passwd, ushort port)
        {
            Host = host;
            Username = userName;
            Password = passwd;
            Port = port;
        }

        public void Run(string type)
        {
            // Console.WriteLine("Prepare {0}... ", type);

            IntPtr conn = NativeMethods.Connect(Host, Username, Password, "", 0);

            if (conn != IntPtr.Zero)
            {
                SQLExe(conn, $"create database if not exists {db} keep 3650");
                SQLExe(conn, $"use {db}");

                if (type == "normal")
                {
                    SQLExe(conn, createStb);
                    SQLExe(conn, createStb1);
                }

                if (type == "json")
                {
                    SQLExe(conn, createJtb);
                    SQLExe(conn, createJtb1);
                }
            }
            else
            {
                throw new Exception("create TD connection failed");
            }

            NativeMethods.Close(conn);
        }

        public void SQLExe(IntPtr conn, string sql)
        {
            IntPtr res;

            res = NativeMethods.Query(conn, sql);
            IfTaosQuerySucc(res, sql);
            NativeMethods.FreeResult(res);
        }

        public bool IfTaosQuerySucc(IntPtr res, string sql)
        {
            if (NativeMethods.ErrorNo(res) == 0)
            {
                return true;
            }
            else
            {
                throw new Exception(
                    $"execute {sql} failed,reason {NativeMethods.Error(res)}, code{NativeMethods.ErrorNo(res)}");
            }
        }
    }
}

File: source/taos-xservice/explorer/taos-connector-node/nodejs/examples/prepare.js
class PrepareData {

    constructor(db, table, numOfRows, ifStable, ts, numOfSub,) {
        this.db = db;
        this.table = table;
        this.numOfRows = numOfRows;
        this.ifStable = ifStable;
        this.ts = ts;
        this.sub = table + "_sub";
        this.numOfSub = numOfSub;
    }

    createDB() {
        return `create database if not exists ${this.db} keep 3650 ;`;
    }
    useDB() {
        return `use ${this.db} ;`;
    }
    createTable() {
        let sql = ""
        if (this.ifStable) {
            sql += `create table if not exists ${this.table} (` +
                ' ts timestamp' +
                ',v1 tinyint' +
                ',v2 smallint' +
                ',v4 int' +
                ',v8 bigint' +
                ',u1 tinyint unsigned' +
                ',u2 smallint unsigned' +
                ',u4 int unsigned' +
                ',u8 bigint unsigned' +
                ',f4 float,f8 double' +
                ',bin binary(200)' +
                ',nchr nchar(200)' +
                ',b bool' +
                ',nilcol int' +
                ')tags(' +
                ' bo bool' +
                ',tt tinyint' +
                ',si smallint' +
                ',ii int' +
                ',bi bigint' +
                ',tu tinyint unsigned' +
                ',su smallint unsigned' +
                ',iu int unsigned' +
                ',bu bigint unsigned' +
                ',ff float' +
                ',dd double' +
                ',bb binary(200)' +
                ',nc nchar(200)' +
                ') ;'

        } else {
            sql += `create table if not exists ${this.table} (` +
                ' ts timestamp' +
                ',v1 tinyint' +
                ',v2 smallint' +
                ',v4 int' +
                ',v8 bigint' +
                ',u1 tinyint unsigned' +
                ',u2 smallint unsigned' +
                ',u4 int unsigned' +
                ',u8 bigint unsigned' +
                ',f4 float,f8 double' +
                ',bin binary(200)' +
                ',nchr nchar(200)' +
                ',b bool' +
                ',nilcol int' +
                ');'
        }
        return sql;
    }

    dropTable() {
        return `drop table if exists ${this.db}.${this.table};`;
    }
    dropDB() {
        return `drop database if exists ${this.db};`;
    }

    insert() {
        if (this.ifStable == false) {
            let str = `insert into ${this.table} values `
            for (let i = 0; i < this.numOfRows; i++) {
                str += '('
                str += `${this.ts}`
                str += `, ${-1 * i + 1}`
                str += `, ${-1 * i + 2}`
                str += `, ${-1 * i + 3}`
                str += `, ${-1 * i + 4}`
                str += `, ${1 * i + 1}`
                str += `, ${1 * i + 2}`
                str += `, ${1 * i + 3}`
                str += `, ${1 * i + 4}`
                str += `, ${3.1415 * i}`
                str += `, ${3.1415926536 * i}`
                str += `, 'binary_col_${i}'`
                str += `, 'nchar_col_${i}'`
                str += `, ${i & 1 == 1 ? true : false}`
                str += ', null'
                str += ')'
                this.ts += 1;
            }
            return str += ";";
        }else{
            throw new Error("using insertStable() for stable");
        }
        
    }

    insertStable() {
        let sql = [];
        for (let i = 0; i < this.numOfSub; i++) {
            let str = `insert into ${this.sub}_${i} using ${this.table} tags`
            str += '('
            str += `  ${i & 1 == 1 ? true : false}`
            str += `, ${-1 * i + 1}`
            str += `, ${-1 * i + 2}`
            str += `, ${-1 * i + 3}`
            str += `, ${-1 * i + 4}`
            str += `, ${1 * i + 1}`
            str += `, ${1 * i + 2}`
            str += `, ${1 * i + 3}`
            str += `, ${1 * i + 4}`
            str += `, ${3.1415 * i}`
            str += `, ${3.1415926536 * i}`
            str += `, 'binary_tag_${i}'`
            str += `, 'nchar_tag_${i}'`
            str += ')values'
            for (let j = 0; j < this.numOfRows; j++) {
                str += '('
                str += `${this.ts}`
                str += `, ${-1 * j + 1}`
                str += `, ${-1 * j + 2}`
                str += `, ${-1 * j + 3}`
                str += `, ${-1 * j + 4}`
                str += `, ${1 * j + 1}`
                str += `, ${1 * j + 2}`
                str += `, ${1 * j + 3}`
                str += `, ${1 * j + 4}`
                str += `, ${3.1415 * j}`
                str += `, ${3.1415926536 * j}`
                str += `, 'binary_col_${j}'`
                str += `, 'nchar_col_${j}'`
                str += `, ${j & 1 == 1 ? true : false}`
                str += ', null'
                str += ')'
                this.ts += 1;
            }
            str += ";"
            sql.push(str);
        }
        return sql;
    }

}

module.exports = PrepareData;